src/Solvers/conjugate_gradient_poisson_solver.jl

```diff
@@ -110,41 +124,72 @@ function compute_preconditioner_rhs!(solver::FourierTridiagonalPoissonSolver, rh
     return nothing
 end
 
-const FFTBasedPreconditioner = Union{FFTBasedPoissonSolver, FourierTridiagonalPoissonSolver}
+struct RegularizedPreconditioner{P, R, D}
+    preconditioner :: P
+    rhs :: R
+    regularization :: D
+end
+
+const SolverWithFFT = Union{FFTBasedPoissonSolver, FourierTridiagonalPoissonSolver}
+const FFTBasedPreconditioner = RegularizedPreconditioner{<:SolverWithFFT}
 
-function precondition!(p, preconditioner::FFTBasedPreconditioner, r, args...)
-    compute_preconditioner_rhs!(preconditioner, r)
-    p = solve!(p, preconditioner)
+function precondition!(p, regularized::FFTBasedPreconditioner, r, args...)
+    solver = regularized.preconditioner
+    compute_preconditioner_rhs!(solver, r)
+    p = solve!(p, solver)
 
+    δ = regularized.regularization
+    rhs = regularized.rhs
     mean_p = mean(p)
+
+    if !isnothing(δ)
+        mean_rhs = mean(rhs)
```

Comment:
In principle, the preconditioner solves the Poisson equation in the underlying grid. Following this logic, it is the mean of `rhs` calculated in the underlying grid that should be subtracted. I have not tested which is better, but I feel the mean in the underlying grid is more consistent mathematically. \(I assume `mean(rhs)` here is the mean of `rhs` calculated in the immersed boundary grid.\)

Reply:
Correct, `mean(rhs)` accounts for the immersed boundary. My objective with this was to define a preconditioner that (approximately) solves the same problem as the CG iteration. This means using `mean` over the whole grid.

Put another way, if we use `mean` over the whole grid, then the mean pressure produced by the preconditioner would be different than the mean pressure that would minimize the conjugate gradient residual?

It's worth testing though. To implement this properly, we need to know the volume of the immersed region which will require an additional computation.